pallets/pallet-asset-swap/src/lib.rs

```diff
@@ -0,0 +1,534 @@
+				log::info!(
+					target: LOG_TARGET,
+					"Failed to convert beneficiary {:?} into v3 `Multilocation` with error {:?}",
+					beneficiary,
+					e
+				);
+				DispatchError::from(Error::<T>::Xcm)
+			})?;
+
+			// 6. Compose and validate XCM message
+			let remote_xcm: Xcm<()> = vec![
+				// TODO: Change this to the configured fee
+				WithdrawAsset((Junctions::Here, 1_000_000_000).into()),
+				SetFeesMode { jit_withdraw: true },
+				TransferAsset {
+					assets: (asset_id_v3, local_asset_amount_as_u128).into(),
+					beneficiary: beneficiary_v3,
+				},
+				// TODO: Add try-catch and asset refund to user account, since we already take them on this chain
+			]
+			.into();
+			let xcm_ticket = validate_send::<T::XcmRouter>(destination_v3, remote_xcm.clone()).map_err(|e| {
+				log::info!(
+					"Failed to call validate_send for destination {:?} and remote XCM {:?} with error {:?}",
+					destination_v3,
+					remote_xcm,
+					e
+				);
+				DispatchError::from(Error::<T>::Xcm)
+			})?;
+
+			// 7. Transfer funds from user to pool
+			let transferred_amount = T::LocalCurrency::transfer(
+				&submitter,
+				&swap_pair.pool_account,
+				local_asset_amount,
+				Preservation::Preserve,
+			)?;
+			if transferred_amount != local_asset_amount {
+				log::error!(
+					"Transferred amount {:?} does not match expected user-specified amount {:?}",
+					transferred_amount,
+					local_asset_amount
+				);
+				return Err(Error::<T>::Internal.into());
+			}
+
+			// 8. Take XCM fee from submitter.
+			let remote_fee_asset_v3: MultiAsset = swap_pair.remote_fee.clone().try_into().map_err(|e| {
+				log::error!(
+					target: LOG_TARGET,
+					"Failed to convert remote fee asset{:?} into v3 `MultiAsset` with error {:?}",
+					swap_pair.remote_fee,
+					e
+				);
+				DispatchError::from(Error::<T>::Internal)
+			})?;
+			let submitter_as_multilocation = T::AccountIdConverter::try_convert(submitter.clone())
+				.map(|j| j.into_location())
+				.map_err(|e| {
+					log::error!(
+						target: LOG_TARGET,
+						"Failed to convert account {:?} into `MultiLocation` with error {:?}",
+						submitter,
+						e
+					);
+					DispatchError::from(Error::<T>::Xcm)
+				})?;
+			let withdrawn_fees =
+				T::AssetTransactor::withdraw_asset(&remote_fee_asset_v3, &submitter_as_multilocation, None).map_err(
+					|e| {
+						log::info!(
+							target: LOG_TARGET,
+							"Failed to withdraw asset {:?} from location {:?} with error {:?}",
+							remote_fee_asset_v3,
+							submitter_as_multilocation,
+							e
+						);
+						DispatchError::from(Error::<T>::UserXcmBalance)
+					},
+				)?;
+			if withdrawn_fees != vec![remote_fee_asset_v3.clone()].into() {
+				log::error!(
+					target: LOG_TARGET,
+					"Withdrawn fees {:?} does not match expected fee {:?}.",
+					withdrawn_fees,
+					remote_fee_asset_v3
+				);
+				return Err(DispatchError::from(Error::<T>::Internal));
+			}
```

Comment:
No. What we withdraw here are the bridged DOTs. So we simply need to "take" them from there, since we will use them on AssetHub. We don't transfer them anywhere on our chain. Is it clear?

Reply:
Sorry Not completely.  
Is the `withdraw_asset` function sending a XCM message to AssetHub, where the `withdrawn_fees` are taken from the sovereign account?

Reply:
No. Here we are withdrawing assets from the user's local DOT balance. Since we know that we will use that on AssetHub to pay for the fees. But the user having that many DOTs means they previously sent them to our sovereign account, so we are using DOTs he previously sent us. So now what we do here is simply saying: "I'm reducing your balance by this much since we will use this much on AssetHub for the wrapped token transfer". It's essentially the same as burning those many DOTs from our local representation.

Reply:
But we already own those DOTs on AssetHub, as they have been reserve-transferred to us alraedy.

Reply:
I see okay